src/redshift/fft/with_precomputation/fft.rs

```diff
@@ -0,0 +1,156 @@
+use crate::pairing::ff::PrimeField;
+use crate::multicore::*;
+use super::FftPrecomputations;
+
+fn log2_floor(num: usize) -> u32 {
+    assert!(num > 0);
+
+    let mut pow = 0;
+
+    while (1 << (pow+1)) <= num {
+        pow += 1;
+    }
+
+    pow
+}
```

Comment:
Saw this same function several times. Perhaps, it is worth moving such math to a specific crate or module.